### PR TITLE
Rename CPU_FREQUENCY_CONTROL to CPU_FREQUENCY_MAX_CONTROL

### LINKED ISSUE #2338: Update MSRIOGroup aliases to provide a one to many alias
> estimate 2

The implementation for the alias feature needs to be modified because it does not allow for a one to many mapping.  Both HWP and the Uncore frequency control aliases would benefit from being able to provide:

- CPU_UNCORE_FREQUENCY_CONTROL --> MSR::UNCORE_RATIO_LIMIT:MIN_RATIO = MSR::UNCORE_RATIO_LIMIT:MAX_RATIO = user request

- CPU_CORE_FREQUENCY_CONTROL --> MSR::HWP_REQUEST:MINIMUM_PERFORMANCE = MSR::HWP_REQUEST:MAXIMUM_PERFORMANCE = user request

Perhaps this can be done as a special feature of the MSRIOGroup, rather than a  general feature of the Control class, in any case, lets make sure the controls are available for 2.0, even if the implementation may change by 2.1.

Done criteria:

- [ ] Integration tests that cover the feature run and pass
- [ ] Dependencies mocked in unit tests are still valid
- [ ] Documentation changes are required since this change modifies interfaces (aliases)
- [ ] Documentation changes are required since this change adds interfaces (uncore aliases)

### PR DESCRIPTION
**Pull request by bgeltz**
- Related #2338.

### CODE REVIEW

**Review by cmcantalupo — DISMISSED** [review]

**Review by dannosliwcd — DISMISSED** [review]

File: service/src/NVMLIOGroup.cpp [hunk 1]
```diff
@@ -689,9 +715,27 @@ namespace geopm
...
+        if (control_name == M_NAME_PREFIX + "GPU_CORE_FREQUENCY_MAX_CONTROL" || control_name == "GPU_CORE_FREQUENCY_MAX_CONTROL") {
+            double min_request;
+            if(!std::isnan(m_frequency_min_control_request.at(domain_idx))) {
+                min_request = m_frequency_min_control_request.at(domain_idx);
+            }
+            else {
+                min_request = m_supported_freq.at(domain_idx).front();
+            }
+            m_nvml_device_pool.frequency_control_sm(domain_idx, min_request, setting / 1e6);
+            m_frequency_min_control_request.at(domain_idx) = setting;
```
Comment (lhlawson) [inline service/src/NVMLIOGroup.cpp]:
```suggestion
            m_frequency_max_control_request.at(domain_idx) = setting;
```

File: service/src/NVMLIOGroup.cpp [hunk 2]
```diff
@@ -689,9 +715,27 @@ namespace geopm
...
+        else if (control_name == M_NAME_PREFIX + "GPU_CORE_FREQUENCY_MIN_CONTROL" || control_name == "GPU_CORE_FREQUENCY_MIN_CONTROL") {
+            double max_request;
+            if(!std::isnan(m_frequency_max_control_request.at(domain_idx))) {
+                max_request = m_frequency_max_control_request.at(domain_idx);
+            }
+            else {
+                max_request = m_supported_freq.at(domain_idx).back();
+            }
+            m_nvml_device_pool.frequency_control_sm(domain_idx, setting / 1e6, max_request);
+            m_frequency_max_control_request.at(domain_idx) = setting;
```
Comment (lhlawson) [inline service/src/NVMLIOGroup.cpp]:
```suggestion
            m_frequency_min_control_request.at(domain_idx) = setting;
```

File: service/src/LevelZeroIOGroup.cpp [hunk 3]
```diff
@@ -818,15 +788,37 @@ namespace geopm
...
-                    write_control(sv.first, sv.second.m_domain_type, domain_idx,
-                                  sv.second.m_controls.at(domain_idx)->m_setting);
+        // If there are any two controls that have an ordering requirement, the retry logic in this
+        // loop enables them to be written in the proper order.
+        // This situation arises due to requirements of minimum settings being less than maximum settings.
+        bool do_throw = false;
+        bool do_retry = false;
+        do {
+            // If the retry succeeds and we do not reset do_retry
+            // then an infinite loop occurs
```
Comment (cmcantalupo) [inline service/src/LevelZeroIOGroup.cpp]:
```suggestion
            // Loop provides one retry attempt.  Setting do_retry to false after the first
            // attempt will end loop after second attempt.
```

File: service/src/LevelZeroIOGroup.cpp [hunk 4]
```diff
@@ -930,28 +922,16 @@ namespace geopm
...
-                restore_control();
-            }
-            else {
-                m_levelzero_device_pool.frequency_control(domain_type, domain_idx,
-                                                          geopm::LevelZero::M_DOMAIN_COMPUTE,
-                                                          setting / 1e6, setting / 1e6);
-            }
-        }
-        else if(control_name == M_NAME_PREFIX + "GPU_CORE_FREQUENCY_MIN_CONTROL") {
+        if(control_name == M_NAME_PREFIX + "GPU_CORE_FREQUENCY_MIN_CONTROL" ||
```
Comment (cmcantalupo) [inline service/src/LevelZeroIOGroup.cpp]:
```suggestion
        if (control_name == M_NAME_PREFIX + "GPU_CORE_FREQUENCY_MIN_CONTROL" ||
```

File: service/src/LevelZeroIOGroup.cpp [hunk 5]
```diff
@@ -930,28 +922,16 @@ namespace geopm
...
+           control_name == "GPU_CORE_FREQUENCY_MIN_CONTROL") {
             double curr_max = read_signal(M_NAME_PREFIX + "GPU_CORE_FREQUENCY_MAX_CONTROL",
                                           domain_type, domain_idx);
             m_levelzero_device_pool.frequency_control(domain_type, domain_idx,
                                                       geopm::LevelZero::M_DOMAIN_COMPUTE,
                                                       setting / 1e6, curr_max / 1e6);
         }
-        else if(control_name == M_NAME_PREFIX + "GPU_CORE_FREQUENCY_MAX_CONTROL") {
+        else if(control_name == M_NAME_PREFIX + "GPU_CORE_FREQUENCY_MAX_CONTROL" ||
```
Comment (cmcantalupo) [inline service/src/LevelZeroIOGroup.cpp]:
```suggestion
        else if (control_name == M_NAME_PREFIX + "GPU_CORE_FREQUENCY_MAX_CONTROL" ||
```

File: service/src/NVMLIOGroup.cpp [hunk 6]
```diff
@@ -548,13 +536,36 @@ namespace geopm
...
-                    write_control(sv.first, sv.second.domain, domain_idx, sv.second.controls.at(domain_idx)->m_setting);
+
+        // If there are any two controls that have an ordering requirement, the retry logic in this
+        // loop enables them to be written in the proper order.
+        // This situation arises due to requirements of minimum settings being less than maximum settings.
+        bool do_throw = false;
+        bool do_retry = false;
+        do {
+            // If the retry succeeds and we do not reset do_retry
+            // then an infinite loop occurs
```
Comment (cmcantalupo) [inline service/src/NVMLIOGroup.cpp]: Update comment here as well.  Unfortunate that this is copy pasted

**Review by cmcantalupo — COMMENTED** [review]
Looks good.  Thanks for all the test updates.  Just a couple of nits (comment and whitespace).

**Review by cmcantalupo — APPROVED** [review]

**Pull request merged**